cdc/split.cc

```diff
@@ -0,0 +1,449 @@
+    // INSERTs can't express updates of individual cells inside a non-atomic
+    // (without deleting the entire field first), so no `nonatomic_updates` field
+    // overwriting a nonatomic column inside an INSERT will be split into two changes:
+    // one with a nonatomic deletion, and one with a nonatomic update
+};
+
+struct clustered_row_update {
+    gc_clock::duration ttl;
+    clustering_key key;
+    std::vector<atomic_column_update> atomic_entries;
+    std::vector<nonatomic_column_deletion> nonatomic_deletions;
+    std::vector<nonatomic_column_update> nonatomic_updates;
+};
+
+struct clustered_row_deletion {
+    clustering_key key;
+    tombstone t;
+};
+
+struct clustered_range_deletion {
+    range_tombstone rt;
+};
+
+struct partition_deletion {
+    tombstone t;
+};
+
+struct batch {
+    std::vector<static_row_update> static_updates;
+    std::vector<clustered_row_insert> clustered_inserts;
+    std::vector<clustered_row_update> clustered_updates;
+    std::vector<clustered_row_deletion> clustered_row_deletions;
+    std::vector<clustered_range_deletion> clustered_range_deletions;
+    std::optional<partition_deletion> partition_deletions;
+};
+
+using set_of_changes = std::map<api::timestamp_type, batch>;
+
+struct row_update {
+    std::vector<atomic_column_update> atomic_entries;
+    std::vector<nonatomic_column_deletion> nonatomic_deletions;
+    std::vector<nonatomic_column_update> nonatomic_updates;
+};
+
+static
+std::map<std::pair<api::timestamp_type, gc_clock::duration>, row_update>
+extract_row_updates(const row& r, column_kind ckind, const schema& schema) {
+    std::map<std::pair<api::timestamp_type, gc_clock::duration>, row_update> result;
+    r.for_each_cell([&] (column_id id, const atomic_cell_or_collection& cell) {
+        auto& cdef = schema.column_at(ckind, id);
+        if (cdef.is_atomic()) {
+            auto view = cell.as_atomic_cell(cdef);
+            auto timestamp_and_ttl = std::pair(
+                    view.timestamp(),
+                    view.is_live_and_has_ttl() ? view.ttl() : gc_clock::duration(0)
+                );
+            result[timestamp_and_ttl].atomic_entries.push_back({id, atomic_cell(*cdef.type, view)});
+            return;
+        }
+
+        cell.as_collection_mutation().with_deserialized(*cdef.type, [&] (collection_mutation_view_description mview) {
+            auto desc = mview.materialize(*cdef.type);
+            for (auto& [k, v]: desc.cells) {
+                auto timestamp_and_ttl = std::pair(
+                        v.timestamp(),
+                        v.is_live_and_has_ttl() ? v.ttl() : gc_clock::duration(0)
+                    );
+                auto& updates = result[timestamp_and_ttl].nonatomic_updates;
+                if (updates.empty() || updates.back().id != id) {
+                    updates.push_back({id, {}});
+                }
+                updates.back().cells.push_back({std::move(k), std::move(v)});
+            }
```

Comment:
Do we really need to care about cell ttls/timestamps? Afaik, we never generate disparate cell timestamps/ttls in any collection code? And the tomstone timestamp should be more or less ignorable also, since it is only a stop marker, guaranteed to be before the cells (which again should be ts:ed to the whole row, no?).

Reply:
> Afaik, we never generate disparate cell timestamps/ttls in any collection code?

We need to handle this somehow:
```
begin batch
update tab using timestamp 42 set coll = coll + {0} where pk = 0 and ck = 0;
update tab using timestamp 43 set coll = coll + {1} where pk = 0 and ck = 0;
apply batch;
```

This will merge the two `collection_mutation`s, so we'll have a `collection_mutation` with cells of different timestamps.

> And the tomstone timestamp should be more or less ignorable also, since it is only a stop marker, guaranteed to be before the cells (which again should be ts:ed to the whole row, no?)

A mutation that represents the following:
```
begin batch
update tab using timestamp 42 set coll = null where pk = 0 and ck = 0;
update tab using timestamp 43 set coll = coll + {0} where pk = 0 and ck = 0;
apply batch;
```
cannot be distinguished from a mutation that represents the following:
```
update tab using timestamp 43 set coll = {0} where pk = 0 and ck = 0
```
so should we represent it in CDC as two changes or one? I decided to go for the first option (two changes), and from the top of my head I can give two reasons:
- there are two different timestamps here, so it will be consistent with the fact that we're splitting different timestamps into different changes everywhere else
- another reason is related to `ttl`s. Again for consistency / to reduce the number of edge cases I decided to group dead cells with `ttl = null`. This is because we now have this `ttl` column which can sometimes be non-null, and it doesn't really make sense to have a change with `ttl != null` and a dead cell, for example, because dead cells cannot have ttls. So the first option above makes even more sense if we do
`
update tab using timestamp 43 and ttl 5 set coll = {0} where pk = 0 and ck = 0
`
because now we'll have two changes: one that sets `coll = null` at timestamp `42` and ttl `null`, and one that sets `coll = coll + {0}` (adds the `0` cell to `coll`) at timestamp `43` with ttl `5`.